middleware/diesel/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+//!     let repo = Repo::new(":memory:");
+//!     // Add the diesel middleware to a new pipeline
+//!     let (chain, pipeline) =
+//!         single_pipeline(new_pipeline().add(DieselMiddleware::new(repo)).build());
+//!
+//!     // Build the router
+//!     build_router(chain, pipeline, |route| {
+//!         route.get("/").to(handler);
+//!     })
+//! }
+//!
+//! fn handler(state: State) -> Box<HandlerFuture> {
+//!     let repo = Repo::borrow_from(&state).clone();
+//!     // As an example, we perform the query:
+//!     // `SELECT 1`
+//!     let f = repo
+//!         .run(move |conn| {
+//!             diesel::select(diesel::dsl::sql("1"))
+//!             .load::<i64>(&conn)
+//!             .map(|v| v.into_iter().next().expect("no results"))
+//!         }).then(|result| match result {
+//!            Ok(n) => {
+//!                let body = format!("result: {}", n);
+//!                let res = create_response(&state, StatusCode::OK, mime::TEXT_PLAIN, body);
+//!                Ok((state, res))
+//!            },
+//!            Err(e) => Err((state, e.into_handler_error())),
+//!         });
+//!    Box::new(f)
+//! }
+//!
+//! # fn main() {
+//! #    let test_server = TestServer::new(router()).unwrap();
+//! #    let response = test_server
+//! #        .client()
+//! #        .get("https://example.com/")
+//! #        .perform()
+//! #        .unwrap();
+//! #    assert_eq!(response.status(), StatusCode::OK);
+//! #    let body = response.read_utf8_body().unwrap();
+//! #    assert_eq!(&body, "result: 1");
+//! # }
+//! ```
+#![doc(test(no_crate_inject, attr(allow(unused_variables), deny(warnings))))]
+
+use diesel::Connection;
+use futures::future::{self, Future};
+use log::{error, trace};
+use std::io;
+use std::panic::{catch_unwind, AssertUnwindSafe};
+use std::process;
+
+use gotham::handler::HandlerFuture;
+use gotham::middleware::{Middleware, NewMiddleware};
+use gotham::state::{request_id, State};
+
+mod repo;
+
+pub use crate::repo::Repo;
+
+/// A Gotham compatible Middleware that manages a pool of Diesel connections via a `Repo` and hands
+/// out connections to other Middleware and Handlers that require them via the Gotham `State`
+/// mechanism.
+pub struct DieselMiddleware<T>
+where
+    T: Connection + 'static,
+{
+    repo: AssertUnwindSafe<Repo<T>>,
+}
+
+impl<T> DieselMiddleware<T>
+where
+    T: Connection,
+{
+    pub fn new(repo: Repo<T>) -> Self {
+        DieselMiddleware {
+            repo: AssertUnwindSafe(repo),
+        }
+    }
+}
+
+impl<T> Clone for DieselMiddleware<T>
+where
+    T: Connection + 'static,
+{
+    fn clone(&self) -> Self {
+        match catch_unwind(|| self.repo.clone()) {
+            Ok(repo) => DieselMiddleware {
+                repo: AssertUnwindSafe(repo),
+            },
+            Err(_) => {
+                error!("PANIC: r2d2::Pool::clone caused a panic");
+                eprintln!("PANIC: r2d2::Pool::clone caused a panic");
```

Comment:
Think you might have left this `eprintln!` accidentally?

Reply:
Indeed, removed.